alf/algorithms/algorithm_interface.py

```diff
@@ -0,0 +1,295 @@
+
+        Args:
+            root_inputs (nest): input for rollout_step() of the root algorithm.
+                This is from replay buffer. Note this is not same as the input
+                of rollout_step() of self unless self is the root algorithm.
+            rollout_info (nested Tensor): ``AlgStep.info`` from rollout_step()
+                for this algorithm.
+            batch_info (BatchInfo): information about this batch of data
+        Returns:
+            tuple:
+            - processed root_inputs
+            - processed rollout_info
+        """
+        return root_inputs, rollout_info
+
+    def after_update(self, root_inputs, info):
+        """Do things after completing one gradient update (i.e. ``update_with_gradient()``).
+        This function can be used for post-processings following one minibatch
+        update, such as copy a training model to a target model in SAC, DQN, etc.
+
+        Args:
+            root_inputs (nest): experiences collected for the most recent
+                ``update_with_gradient()``.
+            info (nest): information collected for training.
+                It is batched from each ``AlgStep.info`` returned by ``rollout_step()``
+                for on-policy training or ``train_step()`` for off-policy training.
+        """
+
+    def after_train_iter(self, root_inputs, rollout_info):
+        """Do things after completing one training iteration (i.e. ``train_iter()``
+        that consists of one or multiple gradient updates). This function can
+        be used for training additional modules that have their own training logic
+        (e.g., on/off-policy, replay buffers, etc). These modules should be added
+        to ``_trainable_attributes_to_ignore`` in the parent algorithm.
+
+        Other things might also be possible as long as they should be done once
+        every training iteration.
+
+        This function will serve the same purpose with ``after_update`` if there
+        is always only one gradient update in each training iteration. Otherwise
+        it's less frequently called than ``after_update``.
+
+        Args:
+            root_inputs (nest): inputs for the rollout_step() of the root algorithm
+                collected during ``unroll()``.
+            rollout_info (nest): information collected from rollout_step() for
+                this algorithm during ``unroll()``.
+        """
+
+    def train_iter(self):
+        """Perform one iteration of training.
+
+        Users may choose to implement their own ``train_iter()``.
+
+        Returns:
+            int:
+            - number of samples being trained on (including duplicates).
+        """
+
+    def train_from_unroll(self, experience, train_info):
+        """Train given the info collected from ``unroll()``. This function can
+        be called by any child algorithm that doesn't have the unroll logic but
+        has a different training logic with its parent (e.g., off-policy).
```

Comment:
I think the meaning of "train_from_unroll" is that when the parent doesn't train from unroll (off-policy) but the child wants to? Otherwise the child can just train together with its parent during unroll.